h2o-genmodel/src/main/java/hex/genmodel/algos/gam/GamMojoModelBase.java

```diff
@@ -0,0 +1,171 @@
+package hex.genmodel.algos.gam;
+
+import hex.genmodel.ConverterFactoryProvidingModel;
+import hex.genmodel.GenModel;
+import hex.genmodel.MojoModel;
+import hex.genmodel.easy.CategoricalEncoder;
+import hex.genmodel.easy.EasyPredictModelWrapper;
+import hex.genmodel.easy.RowData;
+import hex.genmodel.easy.RowToRawDataConverter;
+import hex.genmodel.utils.ArrayUtils;
+import hex.genmodel.utils.DistributionFamily;
+import hex.genmodel.utils.LinkFunctionType;
+
+import java.util.Map;
+import java.util.Set;
+
+public abstract class GamMojoModelBase extends MojoModel implements ConverterFactoryProvidingModel {
+  String _link;
+  public LinkFunctionType _link_function;
+  boolean _useAllFactorLevels;
+  int _cats;
+  int[] _catNAFills;
+  int[] _catOffsets;
+  int _nums;
+  int _numsCenter;
+  double[] _numNAFills;
+  double[] _numNAFillsCenter;
+  boolean _meanImputation;
+  double[] _beta;
+  double[] _beta_no_center;
+  double[] _beta_center;
+  double[][] _beta_multinomial;
+  double[][] _beta_multinomial_no_center; // coefficients not centered for multinomial/ordinal
+  double[][] _beta_multinomial_center; // coefficients not centered for multinomial/ordinal
+  DistributionFamily _family;
+  String[] _gam_columns;
+  int _num_gam_columns;
+  int[] _bs;
+  int[] _num_knots;
+  double[][] _knots;
+  double[][][] _binvD;
+  double[][][] _zTranspose;
+  String[][] _gamColNames;  // expanded gam column names
+  String[][] _gamColNamesCenter;
+  String[] _names_no_centering; // column names of features with no centering
+  int _totFeatureSize; // Gam Algo predictors numbers that include: predictors, expanded gam columns no centered
+  int _betaSizePerClass;
+  int _betaCenterSizePerClass;
+  double _tweedieLinkPower;
+  double[][] _basisVals; // store basis values array for each gam column
+  double[][] _hj;   // difference between knot values
+  int _numExpandedGamCols; // number of expanded gam columns
+  int _lastClass;
+  int _featuresPresentInDataCount = 0; // Count number of predictors found both in input data and Gam Algo predictor
+  
+  public int getTotFeatureSize() { return _totFeatureSize;}
+  public String[] getNames() { return _names; }  // get column names centered
+  public int getFeautesPresentInData() { return _featuresPresentInDataCount;}
+  
+  GamMojoModelBase(String[] columns, String[][] domains, String responseColumn) {
+    super(columns, domains, responseColumn);
+  }
+
+  public void countFeautesPresentInData(Set<String> inputColumnNames) {
```

Comment:
The problem here is that this makes the MOJO stateful. Meaning that two different orders of calls will produce different results.
We definitely need to get rid of _featuresPresentInDataCount.
There are two ways to do this:
1) (which I dont like) is to do countFeautesPresentInData() on every row we want to score, this will make scoring very slow and ultimatelly is not necessary
2) countFeautesPresentInData is present only because testJavaPreditct() is calling this mojo with weird synthetic data, which in turn is caused by the mojo._names containing gam columns, which are synthetic computational columns that should not be in _names in the first place.
WDYT?

Reply:
@honzasterba I think I have a solution for that without using _featuresPresentInDataCount